### PR TITLE
Made it so you do not die on invalid tool

### PR DESCRIPTION
**Pull request by whitead**
^

### CODE REVIEW

File: paperqa/agents/env.py [hunk 1]
```diff
@@ -162,19 +162,31 @@ async def step(
...
-        # If the action has empty tool_calls, the agent can later take that into account
-        msgs = cast(
+        valid_action, invalid_action = self.filter_invalid_tool_calls(action)
+        # NOTE: valid_action or invalid_action may have an empty list of tool calls
+        response_messages = cast(
             list[Message],
-            await self.exec_tool_calls(action, state=self.state, handle_tool_exc=True),
+            # Ordered since things like search -> lookup need to be run in order.
+            # NOTE: Handling tool exceptions here keeps the trajectory going, but I don't
+            # think the returned message is useful to the agent/learning. Disabling for now.
```
Comment (jamesbraza) [inline paperqa/agents/env.py:171]:
```suggestion
```

This comment is for HotPotQA:
- Mentions `lookup` tool
- Doesn't have `handle_tool_exc`

Let's just remove it

File: paperqa/agents/env.py [hunk 2]
```diff
@@ -162,19 +162,31 @@ async def step(
...
+            await self.exec_tool_calls(
+                valid_action, state=self.state, handle_tool_exc=True
+            )
+            + [
+                ToolResponseMessage.from_call(
+                    tool_call, content=f"Invalid tool call: {tool_call.function.name}"
```
Comment (jamesbraza) [inline paperqa/agents/env.py:177]:
```suggestion
                    tool_call, content=f"Tool call with invalid name {tool_call.function.name!r}."
```

Let's be more specific why it's invalid

File: paperqa/agents/env.py [hunk 3]
```diff
@@ -162,19 +162,31 @@ async def step(
...
+                )
+                for tool_call in invalid_action.tool_calls
```
Comment (jamesbraza) [inline paperqa/agents/env.py:179]: One comment on this implementation is the order of `tool_calls` input vs output is not preserved here when there's invalid tool calls

File: paperqa/agents/env.py [hunk 4]
```diff
@@ -162,19 +162,31 @@ async def step(
     ) -> tuple[list[Message], float, bool, bool]:
         self.state.session.add_tokens(action)  # Add usage for action if present
 
-        # If the action has empty tool_calls, the agent can later take that into account
-        msgs = cast(
+        valid_action, invalid_action = self.filter_invalid_tool_calls(action)
```
Comment (jamesbraza) [inline paperqa/agents/env.py:165]: Okay so, based on internal discussions today, I think we should check if `not action.tool_calls`, and return `done` if so

```python
if not action.tool_calls:
    return Message(content="Agent chose to finish."), 0, GenerateAnswer.FAILED_TO_ANSWER, False
```

Comment (jamesbraza) [reply paperqa/agents/env.py:165]: Let's save this for another PR happening after this one, I can take care of it

File: paperqa/agents/env.py (hunk 3 above)
Comment (jamesbraza) [reply paperqa/agents/env.py:179]: We can save this for another PR too

File: paperqa/agents/env.py (hunk 4 above)
Comment (jamesbraza) [reply paperqa/agents/env.py:165]: Handled this separately in https://github.com/Future-House/paper-qa/pull/671

File: pyproject.toml [hunk 5]
```diff
@@ -42,13 +42,12 @@ dependencies = [
...
     "anyio",
     "coredis",
-    "fhaviary[llm]>=0.8.2",  # For namespace reorg
+    "fhaviary[llm]>=0.10.0",  # For invalid tool fixes
     "html2text",  # TODO: evaluate moving to an opt-in dependency
     "httpx",
     "limits",
     "litellm>=1.44",  # to prevent sys.stdout on router creation
     "numpy",
-    "openai<1.47",  # TODO: remove after https://github.com/BerriAI/litellm/issues/5854 resolution
```
Comment (jamesbraza) [inline pyproject.toml:51]: @whitead can you remove this from `pre-commit`'s `mypy` as well?

File: pyproject.toml [hunk 6]
```diff
@@ -42,13 +42,12 @@ dependencies = [
     "aiohttp",  # TODO: remove in favor of httpx
     "anyio",
     "coredis",
-    "fhaviary[llm]>=0.8.2",  # For namespace reorg
```
Comment (jamesbraza) [inline pyproject.toml:45]: Can you match this in `pre-commit`'s `mypy`

File: paperqa/agents/env.py [hunk 7]
```diff
@@ -174,19 +174,23 @@ async def step(
...
-        # If the action has empty tool_calls, the agent can later take that into account
-        msgs = cast(
+        response_messages = cast(
             list[Message],
-            await self.exec_tool_calls(action, state=self.state, handle_tool_exc=True),
+            await self.exec_tool_calls(
+                action,
+                state=self.state,
+                handle_tool_exc=True,
+                ordered=True,  # prevent concurrent gather evidence steps or anything
```
Comment (jamesbraza) [inline paperqa/agents/env.py:183]: Can you do `ordered=True` in another PR?

Alternately, can you put a `TODO` here to remove this?

File: pyproject.toml (hunk 5 above)
Comment (whitead) [reply pyproject.toml:51]: Isn't in pre-commit

File: pyproject.toml (hunk 6 above)
Comment (whitead) [reply pyproject.toml:45]: good catch

File: paperqa/agents/env.py (hunk 7 above)
Comment (whitead) [reply paperqa/agents/env.py:183]: Do you think it's not a bug to allow concurrent gather evidence?

Comment (whitead) [reply paperqa/agents/env.py]: removed

File: .pre-commit-config.yaml [hunk 8]
```diff
@@ -85,7 +85,7 @@ repos:
         additional_dependencies:
           - aiohttp
           - coredis
-          - fhaviary[llm]>=0.8.2 # Match pyproject.toml
+          - fhaviary[llm]>=0.10.0 # Match pyproject.toml
```
Comment (jamesbraza) [inline .pre-commit-config.yaml:88]: A few line below, let's remove the:

```toml
- openai>=1 # Match pyproject.toml
```

File: pyproject.toml (hunk 5 above)
Comment (jamesbraza) [reply pyproject.toml:51]: Are you sure

File: .pre-commit-config.yaml (hunk 8 above)
Comment (whitead) [reply .pre-commit-config.yaml:88]: Do you not want to have openai type checking?

Comment (jamesbraza) [reply .pre-commit-config.yaml:88]: So `mypy` will still get `openai` through `litellm`, as `openai` is a `litellm` dependency. So we'll still have `openai` there.

The idea is we want the `mypy` list to match our first party deps in `dependencies`.

We made `openai` a first party dependency temporarily just to ensure its version matched. So we had to do the same in `mypy`. Now that this PR removes it from our first party deps, we want to remove it from our `mypy` list as well to keep them in sync

Comment (whitead) [reply .pre-commit-config.yaml:88]: Got it - this section of mypy has many inconsistencies with the pyproject.toml file. Will fix it I guess

Comment (whitead) [reply .pre-commit-config.yaml:88]: Like `litellm` isn't even in the `mypy` file

Comment (jamesbraza) [reply .pre-commit-config.yaml:90]: Yes you are dead on 🥳 

One last dimension (we also just talked in person) is that some packages are untyped: https://github.com/Future-House/paper-qa/blob/v5.3.2/pyproject.toml#L144-L149, so we don't need to give them to `mypy`'s isolated list here. Respecifying them (even though they aren't typed):
- Pro: future looking for if they become typed
- Con: bloats `pre-commit`'s isolated venv a bit

That being said, it's an easier heuristic to just "exact match" dependency lists. So let's leave as is, thanks and resolving

**Review by jamesbraza — APPROVED** [review]
LGTM, thanks for doing this!

File: pyproject.toml [hunk 9]
```diff
@@ -258,6 +257,8 @@ min-similarity-lines = 12
 # Add the specified OPTS to the set of command line arguments as if they had been
 # specified by the user.
 addopts = "--typeguard-packages=paperqa"
+# set how loops are scoped to avoid https://github.com/BerriAI/litellm/issues/5854
+asyncio_default_fixture_loop_scope = "session"
```
Comment (jamesbraza) [inline pyproject.toml:261]: Confirmed this worked in https://github.com/Future-House/paper-qa/pull/674

Nice work here @whitead, we are unblocked again

**Pull request merged**